src/lib/asl.c

```diff
@@ -0,0 +1,56 @@
+#include <asl.h>
+#include <pwd.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <time.h>
+
+#include <SystemConfiguration/SystemConfiguration.h>
+
+#include <xhyve/asl.h>
+
+static aslclient asl = NULL;
+static aslmsg log_msg = NULL;
+
+static unsigned char *buf = NULL;
+static size_t buf_size = 0;
+static size_t buf_capacity = 0;
+
+/* asl is deprecated in favor of os_log starting with macOS 10.12.  */
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+
+/* Grow buf/buf_capacity. */
+static void buf_grow(void)
+{
+	buf_capacity = buf_capacity ? 2 * buf_capacity : 1024;
+	buf = realloc(buf, buf_capacity);
+	if (!buf) {
+		perror("buf_grow");
+		exit(1);
+	}
+}
+
+
+/* Initialize ASL logger and local buffer. */
+void asl_init(void)
+{
+	asl = asl_open(NULL, NULL, 0);
+	log_msg = asl_new(ASL_TYPE_MSG);
+	buf_grow();
+}
+
+
+/* Send one character to the logger: wait for full lines before actually sending. */
+void asl_put(uint8_t c)
+{
+	if (buf_size + 1 >= buf_capacity) {
+		buf_grow();
+	}
+	if (c == '\n') {
```

Comment:
I'm slightly concerned about what happens if the console is e.g. running some sort of curses app or something where `\n` might conceivably be quite rare, it could end up building up quite a large buffer.

Perhaps add an upper bound on the buffer and flush then even if no `\n` is observed?

Also don't you need to handle `\0` specially here? Without such handling the buffer passed to `asl_log` will be truncated at any NUL in the stream. Perhaps just treat `\0` as equivalent to `\n`?

Reply:
I have no idea what can be sent to tty here, whatever you suggest, I'll buy!  I'm fine with your \0 = \n approach, or `\0` = `\\0` to make it visible, etc.

If you are aware of a tty2text approach, I'll take it too.

Reply:
I think what can be sent to the console here is effectively arbitrary bytes, could be anything, from raw binary to utf-16 encoded text, there's probably no good universal answer. Most frequently it'll be ASCII or UTF-8 though plus some escape codes, I expect.
 
How about:
* Put an upper limit of 4K on the buffer. If we reach this limit then flush unconditionally (could also make it a static 4K buffer if you like since the dynamism is no longer that important with a static limit in place).
* Treat `\n` and `\0` the same: do not inject them in the log stream, instead flush the current buffer.
* For `\` insert it as `\\` (i.e. escape it, to distinguish from the result of the next rule)
* If `!isprint(c)` then escape it as hex `\xx`
* otherwise simply insert `c` as is.

This should preserve all of the useful text in the common case without injecting control characters or anything too weird into ASL (which I don't expect to handle anything especially crazy). It's not quite reversible since we've lost the distinction between `\n` and `\0` but I think it'll do, including the `\n` would make things pretty unreadable in the common case.

The escaping makes the buffer size check a little more complicated, since it's no longer `1:1`. I suggest just growing when you reach `capacity - 3` rather than being too fancy.

Reply:
If we put an upper limit of 4k on the buffer, why not just statically allocate it?

Reply:
Looking at http://www.manpagez.com/man/3/asl_log/ I'm half wondering if perhaps ASL is already setup to deal with such things correctly. There's a bunch of `ASL_ENCODING_*` which look to only be used with `asl_add_output_file` but there seems to be a general implication that the internal DB treats things as NUL terminated strings ("All keys and values are NUL-character terminated C language strings.  UTF-8 encoding may be used for non-ASCII characters.") and it understands escaping in the presentation layer to some extent.

So, while I'm a bit wary that we're going to bring things crashing down around us lets try just flushing on a) 4K limit and b) on `\n` or `\0` and hope for the best.